Added fetch look back docs (#1087)
demisto/content-docs: docs/integrations/fetching-incidents.md

During a ***fetch-incidents*** run, some edge-case scenarios may cause missing incidents from the 3rd-party product.  
The most common scenarios are:
* Indexing issues in the product: For example, if incident A was created before incident B and only B was indexed, ***fetch-incidents*** will fetch only B and not A. If A was indexed after the fetch was called, the next fetch will fetch only from the created time of B.
* An update in the incident information: Some implementations of ***fetch-incidents*** use a query filter to fetch only specific incidents. If initially an incident did not match the query (meaning, it was not fetched) but at some point was updated so that it now matches the query, ***fetch-incidents*** will not pull the updated incident because the time to fetch it already passed.

### Solution
The *look_back* parameter enables configuring how far back in time (in minutes) ***fetch-incidents*** will look to get the incidents that were created a while ago but indexed a few minutes ago.  
In addition, the **LastRun** object stores the following fields to be used by the lookback methods:
- **time** - The time to fetch the next fetch call (as in a regular fetch).

> - **limit** - The number of incidents retrieved in the next fetch. If the current fetch run has the same start_time as the last fetch (determined in **get_fetch_run_time_range()**), the limit will be increased by the limit parameter value, and then incidents retrieved in the last fetch will be filtered out.

Reviewer: ```suggestion
- **limit** - The maximal number of incidents retrieved in the next fetch. If the current fetch run has the same start_time as the last fetch (determined in **get_fetch_run_time_range()**), this field will be increased by the limit instance parameter value, and then incidents retrieved in the last fetch will be filtered out.
```
just to make it easier to follow (to avoid confusion between the last run field and the parameter)

    query = build_query(start_fetch_time, end_fetch_time, fetch_limit)
    incidents_res = get_incidents_request(query)

    incidents = filter_incidents_by_duplicates_and_limit(incidents_res=incidents_res, last_run=last_run,
                                                         fetch_limit=fetch_limit_param, id_field='incident_id')

    last_run = update_last_run_object(last_run=last_run, incidents=incidents, fetch_limit=fetch_limit_param,
                                      start_fetch_time=start_fetch_time, end_fetch_time=end_fetch_time, look_back=look_back, 
                                      created_time_field='created', id_field='incident_id')

    demisto.incidents(incidents)
    demisto.setLastRun(last_run)
```

### Notes:
- Fetching incidents is flexible, you can use the various functions according to your needs.
- You can also use the generic methods for regular ***fetch-incidents*** without lookback.
- If the *look_back* value is increased by *k* minutes, you may get duplicate incidents for the *k* minutes that overlap with the previous fetch.

## Troubleshooting

> To troubleshoot ***fetch-incident***, execute `!integration_instance_name-fetch` in the Playground to return the incidents.  

Reviewer: ```suggestion
To troubleshoot ***fetch-incident***, execute `!integration_instance_name-fetch debug-mode=true` in the Playground to return the incidents.  
```

demisto.incidents([])
```

## Advenced

### Generic fetch with look back
Use the generic look back functions for fetching missing incidents that are indexed a while after they are created.

#### Use case
When a 3rd party product creates incidents, sometimes it might be indexed a while after it was created.
For example, if incident "a" was created before incident "b", and just "b" was indexed, the fetch-incidents will fetch only "b" and will miss "a", also if "a" was indexed after the fetch were called, the next fetch will fetch only from the created time of "b".

#### Solution
Having the parameter look_back we can configure how nuch the fetch-incidents will look back in time (minutes) to get the incidents that were created a while ago but indexed a minute ago.

The solution functions are in [CommonServerPython](https://xsoar.pan.dev/docs/reference/api/common-server-python) and you can find the exact documentation there, but here is a short description.

> * **get_fetch_run_time_range** - The function gets the last run object and some other arguments and calculate the time to fetch and returns the start and end time to fetch.

Reviewer: ```suggestion
* **get_fetch_run_time_range()** - using the last run object and other necessary parameters, this method calculates and retrieves the time range from which to fetch.
```

> * **get_fetch_run_time_range()** - using the last run object and other necessary parameters, this method calculates and retrieves the time range from which to fetch.

Reviewer 2: make sure to state that if use that for fetching, you must always get the ```end_fetch_time``` as well

Author: I mentioned this in the code example.

If the parameter look_back is given, then the start time will always be >= than `now - look_back`.

* **filter_incidents_by_duplicates_and_limit()** - After getting the incidents from the 3rd party API call, we need to filter out the duplicate incidents.

> In the example above, after an incident A is indexed the fetch will bring incidents A and B, then we must return only A because B was already fetched. In addition, after filtering duplicates, if we still have more incidents than the limit, the function will return only by the limit.

Reviewer 2: ```suggestion
In the example below, after an incident A is indexed the fetch will bring incidents A and B, then we must return only A because B was already fetched. In addition, after filtering duplicates, if we still have more incidents than the limit, the function will return only by the limit.
```

don't you mean the example below (python code)?

Author: I meant the example in line 134.

> * **filter_incidents_by_duplicates_and_limit** - After getting the incidents from the 3rd party API call, we will want to filter the incidents from duplicates.
> In the example above, after incident "a" is indexed the fetch will bring incidents "a" and "b", then we want to return only "a" because "b" is already fetched.
> In addition, after filtering from duplicates, if we still have more incidents than the limit, the function will return only by the limit.
> 
> * **get_latest_incident_created_time** - Given a list of incidents and the created time field, the function will return the latest incident created time.
> 
> * **remove_old_incidents_ids** - Removes old unneccesary incident ids from the last run object to avoid overloading.
> 
> * **get_found_incident_ids** - Returns a list of the new fetched incident ids. This is for saving it into the last run object and filter duplicates in the next call of fetch-incidents.
> 
> * **create_updated_last_run_object** - Creates a new last run object with a new time and limit for the next fetch.
> 
> * **update_last_run_object** - Updates the existing last run object.

Reviewer: ```suggestion
* **filter_incidents_by_duplicates_and_limit()** - After getting the incidents from the 3rd party API call, we need to filter out the duplicate incidents.
In the example above, after an incident A is indexed the fetch will bring incidents A and B, then we must return only A because B was already fetched. In addition, after filtering duplicates, if we still have more incidents than the limit, the function will return only by the limit.

* **get_latest_incident_created_time()** - Given a list of incidents and the created time field, the function will return the latest incident created time.

* **remove_old_incidents_ids()** - Removes old incident IDs from the last run object to avoid overloading.

* **get_found_incident_ids()** - Returns a list of the new fetched incident IDs. This is for saving it into the last run object and filter duplicates in the next call of fetch-incidents.

* **create_updated_last_run_object()** - Creates a new last run object with a new time and limit for the next fetch.

* **update_last_run_object()** - Updates the existing last run object.
```

The function updates the found ids given from the function `get_found_incident_ids` and updates also the new time and limit given from the function `create_updated_last_run_object` and returns the updated last run object.

The main way how it works is by saving into the last run object the following fields:
* **time** - As in a regular fetch, the time to fetch from in the next fetch call.
* **limit** - The limit will be increased in case we didn't fetch all the incidents from a given time, so we will save the start time for the next fetch as the current fetch start time.
* **found_incident_ids** - The new fetched incident IDs. This is for filter duplicates in the next fetch calles.

#### Notes
- The solution is highly modular, this is done in part so you can only use specific functions and implement the others according to your needs.
- The generic methods can be used also for regular fetch-incidents without look back as well (where `look_back=0`).
- If the look_back config parameter is inceased for some reason, you might have duplicate incidents as their IDs are not saved in the last run any more and could not be filtered.

### Example fetch with look back
```python

> def fetch_incidents(params):

Reviewer: please format the code according to our conventions (pep8, typing, etc)

demisto.incidents([])
```

## Advenced

### Generic fetch with look back
Use the generic look-back methods for fetching missing incidents.

#### Use case
During a **fetch-incidents** run, some edge-cases might cause missing incidents from the 3rd-party product. The most common are:
* Indexing issues in the product: for example, if an incident A was created before an incident B, and only B was indexed, the fetch-incidents will fetch only B and will miss A, so if A was indexed after the fetch was called, the next fetch will fetch only from the created time of B.
* An update in the incident information: some implementations of fetch-incidents are using a query filter to fetch only specific incidents. If initially an incident did not match the query (i.e., it was not fetched) but at some point was updated so that now it does match the query, a **fetch-incidents** command that queries incidents from the product using a **time** field, will not find it.

#### Solution
Having the *look_back* parameter, we can configure how much the fetch-incidents will look back in time (minutes) to get the incidents that were created a while ago but indexed a minute ago.

> As part of the feature, the following fields are stored in the last run object and used in the methods described below:
> * **time** - As in a regular fetch, the time to fetch from in the next fetch call.
> * **limit** - The limit will be increased by the limit parameter value in case we will need to do the next fetch from the same time we did the current fetch, this is so that we do not get the same incidents result as we got in the current fetch but we will get more incidents and filter out the duplicates.
> * **found_incident_ids** - IDs of incidents fetched in previous runs. Used for filtering duplicates in the next runs.
> 
> The generic methods for this feature are implemented in [CommonServerPython](https://xsoar.pan.dev/docs/reference/api/common-server-python) and you can find the exact documentation there, but here is a short description.
> 
> * **get_fetch_run_time_range()** - using the last run object and other necessary parameters, this method calculates and retrieves the time range from which to fetch.
> If the parameter look_back is given, then the start time will always be >= than `now - look_back`.
> 
> * **filter_incidents_by_duplicates_and_limit()** - After getting the incidents from the 3rd party API call, we need to filter out the duplicate incidents.
> In the example above, after an incident A is indexed the fetch will bring incidents A and B, then we must return only A because B was already fetched. In addition, after filtering duplicates, if we still have more incidents than the limit, the function will return only by the limit.
> 
> * **get_latest_incident_created_time()** - Given a list of incidents and the created time field, the function will return the latest incident created time.
> 
> * **remove_old_incidents_ids()** - Removes old incident IDs from the last run object to avoid overloading.
> 
> * **get_found_incident_ids()** - Returns a list of the new fetched incident IDs. This is for saving it into the last run object and filter duplicates in the next call of fetch-incidents.
> 
> * **create_updated_last_run_object()** - Creates a new last run object with a new time and limit for the next fetch.
> 
> * **update_last_run_object()** - Updates the existing last run object.
> The function updates the found ids given from the function `get_found_incident_ids` and updates also the new time and limit given from the function `create_updated_last_run_object` and returns the updated last run object.

Reviewer: a minor change in `limit` description, also reordered/titled the sections to ease the understanding of what we are going to use eventually.
```suggestion
##### Related LastRun Fields
As part of the feature, the following fields are stored in the last run object and used in the methods described below:
* **time** - As in a regular fetch, the time to fetch from in the next fetch call.
* **limit** - In case the current fetch run has the same start_time we did the last fetch (determined in `get_fetch_run_time_range()`), the limit will be increased by the limit parameter value, and then incidents we already got in the last fetches will be filtered out.
* **found_incident_ids** - IDs of incidents fetched in previous runs. Used for filtering duplicates in the next runs.

##### Related Methods
The generic methods for this feature are implemented in [CommonServerPython](https://xsoar.pan.dev/docs/reference/api/common-server-python) and you can find the full documentation there.

* **get_fetch_run_time_range()** - using the last run object and other necessary parameters, this method calculates and retrieves the time range from which to fetch.
If the parameter look_back is given, then the start time will always be >= than `now - look_back`.

* **filter_incidents_by_duplicates_and_limit()** - After getting the incidents from the 3rd party API call, we need to filter out the duplicate incidents.
In the example above, after an incident A is indexed the fetch will bring incidents A and B, then we must return only A because B was already fetched. In addition, after filtering duplicates, if we still have more incidents than the limit, the function will return only by the limit.

* **update_last_run_object()** - Updates the existing last run object.
The function updates the found ids given from the function `get_found_incident_ids` and updates also the new time and limit given from the function `create_updated_last_run_object` and returns the updated last run object.

###### Helper Methods
The following helpers are used in the methods above, and should not be used at the integration scope.

* **get_latest_incident_created_time()** - Given a list of incidents and the created time field, the function will return the latest incident created time.

* **remove_old_incidents_ids()** - Removes old incident IDs from the last run object to avoid overloading.

* **get_found_incident_ids()** - Returns a list of the new fetched incident IDs. This is for saving it into the last run object and filter duplicates in the next call of fetch-incidents.

* **create_updated_last_run_object()** - Creates a new last run object with a new time and limit for the next fetch.
```
